Update documentation for v32

AyanSinhaMahapatra opened the pull request:
Fixes #3277 
Fixes #3278 
Fixes #3298

### Tasks

* [x] Reviewed [contribution guidelines](https://github.com/nexB/scancode-toolkit/blob/develop/CONTRIBUTING.rst)
* [x] PR is descriptively titled 📑 and links the original issue above 🔗
* [x] Tests pass -- look for a green checkbox ✔️ a few minutes after opening your PR
  Run [tests](https://scancode-toolkit.readthedocs.io/en/latest/contribute/contrib_dev.html#running-tests) locally to check for errors. 
* [x] Commits are in uniquely-named feature branch and has no merge conflicts 📁

AyanSinhaMahapatra committed:
Fix unordered lists bug

Looks like there was an issue in docutils/sphinx-rtd-theme so we
needed to downgrade to docutils < 0.17 for the unordered lists to
be rendered again properly.

Reference: https://github.com/readthedocs/sphinx_rtd_theme/issues/1115
Signed-off-by: Ayan Sinha Mahapatra <ayansmahapatra@gmail.com>

AyanSinhaMahapatra committed:
Add docs about M1 support #3277

Signed-off-by: Ayan Sinha Mahapatra <ayansmahapatra@gmail.com>

AyanSinhaMahapatra committed:
Update How-to-Guides with license updates

Signed-off-by: Ayan Sinha Mahapatra <ayansmahapatra@gmail.com>

AyanSinhaMahapatra committed:
Update license detection reference documentation

Refernce: https://github.com/nexB/scancode-toolkit/issues/3278
Signed-off-by: Ayan Sinha Mahapatra <ayansmahapatra@gmail.com>

AyanSinhaMahapatra committed:
Misc doc changes and test fixes

Signed-off-by: Ayan Sinha Mahapatra <ayansmahapatra@gmail.com>

AyanSinhaMahapatra committed:
Update contribution and other misc docs

Signed-off-by: Ayan Sinha Mahapatra <ayansmahapatra@gmail.com>

AyanSinhaMahapatra committed:
Add auto generated doc page on available package parsers

- Add new console script `regen-package-docs`
- Add new shell script `regen_package_docs.sh`
- Add new jinja template to generate docs using data
- Add this script on makefile to run automatically on doc generation
- Move explanations -> reference section

Reference: https://github.com/nexB/scancode-toolkit/issues/3298
Signed-off-by: Ayan Sinha Mahapatra <ayansmahapatra@gmail.com>

AyanSinhaMahapatra committed:
Misc doc updates

- Add doc check reminder in PR template
- Delete stray file
- Misc license related updates to align docs

Signed-off-by: Ayan Sinha Mahapatra <ayansmahapatra@gmail.com>

AyanSinhaMahapatra committed:
Update the available package parsers table

Updating the column order in available package
parsers from feedback.

Signed-off-by: Ayan Sinha Mahapatra <ayansmahapatra@gmail.com>

AyanSinhaMahapatra committed:
Update scancode CLI reference docs for v32

Signed-off-by: Ayan Sinha Mahapatra <ayansmahapatra@gmail.com>

AyanSinhaMahapatra committed:
Update README.rst

Signed-off-by: Ayan Sinha Mahapatra <ayansmahapatra@gmail.com>

AyanSinhaMahapatra committed:
Merge branch 'develop' into doc-update-licenses

Signed-off-by: Ayan Sinha Mahapatra <ayansmahapatra@gmail.com>

pombredanne committed:
Fix minor typo in documentation

Signed-off-by: Philippe Ombredanne <pombredanne@nexb.com>

AyanSinhaMahapatra commented:
Btw, doc builds for this branch are enabled and can be viewed here: https://scancode-toolkit.readthedocs.io/en/doc-update-licenses/

johnmhoran commented on README.rst:
(As agreed I am only vetting changes, including only those links inside changed passages.)

Line 16: "visualization of scancode results data" -- when do we want to refer to ScanCode vs. scancode?

Lines 73-74: the link intended to be a link to the SCWB GH repo is not working properly -- I think we have an extra space before the closing `>`.

Line 97: this link takes us to a 404 error page:

See `all available package parsers <https://scancode-toolkit.readthedocs.io/en/stable/reference/available_package_parsers.html>`_

Line 168: "After scancode is installed" -- do we want instead to refer to `ScanCode`?  `ScanCodeToolkit`?  `SCTK`?  Note that above we also refer to the "scancode-workbench project", but on that project's readme we call it `ScanCode Workbench` and there we refer to SCTK as `ScanCode`.  See https://github.com/nexB/scancode-workbench  (I won't keep repeating this comment re consistent/desire naming conventions.)

Line 218: this link takes us to "This site can't be reached":

or an `IRC bridge <https://irc.gitter.im/>`_.

AyanSinhaMahapatra replied:
> available_package_parsers.html

This will be available when we merge the PR, all other comments addressed.

johnmhoran commented on docs/source/cli-reference/basic-options.rst:
docs/source/cli-reference/basic-options.rst

Lines 89-90:  "If we run license detection (with --license-text) on the above text"

Perhaps put "--license-text" inside backticks?  In my view of the RTD page, that flag is displayed with a single hyphen/dash, "-license-text", even though when copied and pasted from the RTD page to, e.g., Notepad+ I see a double-dash.

Line 124: change "occurances" to "occurrences"

Lines 669-670:  In the code we have these 2 URLs, which look like markup inside this .rst file:

        [`WTFPL`](http://www.wtfpl.net/txt/copying/) and
        [`MIT`](https://opensource.org/licenses/MIT).
		
and in the RTD they show up like this (no idea how this will be displayed in my comment):

```
        [`WTFPL`](http://www.wtfpl.net/txt/copying/) and
        [`MIT`](https://opensource.org/licenses/MIT).
```

Maybe we want this instead?

```
        `WTFPL <http://www.wtfpl.net/txt/copying/>`_ and
        `MIT <https://opensource.org/licenses/MIT>`_.
```

Line 673:  In "we have to following license detection result" replace "to" with "the"

Lines 748-752:  In this passage:

```
    Here from the ``"detection_log": ["unknown-intro-followed-by-match"]`` added diagnostics
    information we get to know that there was a unknown intro license match, followed by
    proper detections, so we concluded the unknown intro to be a introduction to the
    following license and hence concluded the license from the license matches after the
    unknown detection.
```

consider
- replacing "get to know" with "learn"
- replacing "a unknown" with "an unknown"
- replacing "so we concluded" with "so we conclude"
- replacing "a introduction" with "an introduction"
- replacing "and hence concluded" with "and hence conclude"

AyanSinhaMahapatra replied:
> [`WTFPL`](http://www.wtfpl.net/txt/copying/)

This is just an example file where this was markdown, we did not want a link here. So this is ok.
All other comments addressed.

johnmhoran commented on docs/source/cli-reference/help-text-options.rst:
docs/source/cli-reference/help-text-options.rst

Lines 52-70: The first line of the definitions is indented one space too far, e.g.,

```
      --license-diagnostics        In license detections, include diagnostic details
                                  to figure out the license detection post
                                  processing steps applied.
```

(The indent issue also applies throughout this .rst page.)

In addition, the code-block in which this is displayed, defined by the `::` at the end of the intro line, is an incorrect style -- the `and` and `or` conjunctions (and other words, e.g., `for`) are highlighted in purple and we do not want any highlighting.  See https://scancode-toolkit.readthedocs.io/en/doc-update-licenses/cli-reference/help-text-options.html  

We want something like this to start the code block:

```
.. code-block:: none 

    Blah and blah or blah for blah
```

Line 92: remove extra space before closing `)`

Lines 121-122:

Consider replacing

```
Classify files with flags telling if the file is a
legal, or readme or test file, etc.
```

with something like

```
Classify files with flags indicating whether the file is a
legal, readme, test or similar file
```

Line 140:

Replace `rulespresent in detections.` with `rules present in detections.`

Line 206: "Note: when you run scancode" -- do we mean "ScanCode-Toolkit"?  Do we care about consistent product names?

AyanSinhaMahapatra replied:
All comments addressed. Thanks!

johnmhoran commented on docs/source/cli-reference/scan-options-post.rst:
docs/source/cli-reference/scan-options-post.rst

Line 32: Change 'Their' to 'their' in `Then, the following directories are marked as "Source", i.e. Their "is_source" attribute is set`

Line 33: Change `to "True", as they contain source code majorly.` to `to "True", as they contain mostly source code.`

Line 56:  In `scancode-toolkit as top level packages, dependenices and licenses` change `dependenices` to `dependencies`

Line 57: Change `now provide a improved consolidated data.` to `now provide improved consolidated data.`

Line 206: `such as the ScanCode LicenseDB` -- is that the current name of the DB?

Line 252:  What does this mean?

`It whole JSON file is structured as follows, when it has "license_clarity_score"::`

Does this capture the intent?

`When the "license_clarity_score" is included, the entire JSON file is structured as follows::`

Line 349: change `seperately` to `separately`

Line 479: change `comparision` to `comparison`

Line 600: change `dependenices` to `dependencies`

AyanSinhaMahapatra replied:
> such as the ScanCode LicenseDB -- is that the current name of the DB?

Yes! See https://github.com/nexB/scancode-licensedb#scancode-licensedb and https://scancode-licensedb.aboutcode.org/

Addressed all comments otherwise.

johnmhoran commented on docs/source/cli-reference/scan-options-pre.rst:
docs/source/cli-reference/scan-options-pre.rst

Line 183: In `A facet is essentially file purpose classification label.` add `a` before `file`.

AyanSinhaMahapatra replied:
Addressed, thanks!

johnmhoran commented on docs/source/contribute/contrib_dev.rst:
docs/source/contribute/contrib_dev.rst

Lines 198-199:  These were not edited but they were included in the GH excerpts and once I saw them, well, you know . . .

In `ScanCode can be used alright as a Python library and is available as as a Python wheel in Pypi` delete one `as` and change `Pypi` to `PyPI`.

Please also delete `alright`.

AyanSinhaMahapatra replied:
Thanks, addressed.

johnmhoran commented on docs/source/contribute/contrib_doc.rst:
docs/source/contribute/contrib_doc.rst

Lines 319-323: Each of these 3 links takes me to a 404 page:

It's possible to generate docs automatically from data by using a combination of:

- `shell scripts: example <https://github.com/nexB/scancode-toolkit/blob/develop/docs/scripts/regen_package_docs.sh>`_
- `python scripts: example <https://github.com/nexB/scancode-toolkit/blob/develop/src/packagedcode/regen_package_docs.py>`_
- `jinja templates: example <https://github.com/nexB/scancode-toolkit/blob/develop/src/packagedcode/templates/available_package_parsers.rst>`_

AyanSinhaMahapatra replied:
Yup, these would be working when we merge this PR :P As they are in this PR actually.

johnmhoran commented on docs/source/contribute/cut_new_release.rst:
docs/source/contribute/cut_new_release.rst

Line 1: Delete the colon in the title `How to cut a new release:`

Lines 22-23: In `pull Request` capitalize the `P` in `pull`.

Line 34: In `Merge this ``release-prep-v1.6.1`` branch in master after review approval`, change `master` to `main` (or whatever else has replaced `master`).

-- this likely appears throughout this and other documentation -- I am not looking for it and will not mention it again in these comments.

Line 44: In `We have a `automated release script <https://github.com/nexB/scancode-toolkit/actions/workflows/scancode-release.yml>`_`, change `a` to `an`.

Line 52: change `arcives` to `archives`

AyanSinhaMahapatra replied:
Thanks, addressed.

johnmhoran commented on docs/source/contribute/roadmap.rst:
docs/source/contribute/roadmap.rst

Line 13: In "see the `milstones" change "milstones" to "milestones".

AyanSinhaMahapatra replied:
Addressed, thanks!

johnmhoran commented on docs/source/getting-started/install.rst:
docs/source/getting-started/install.rst

Line 347: In `some non-native dependencies does not have pre-built wheels` changes `does` to `do`.

AyanSinhaMahapatra replied:
Addressed, thanks!

johnmhoran commented on docs/source/getting-started/newcomer.rst:
docs/source/getting-started/newcomer.rst

Line 65: change `then visualize the results.` to `and then visualizing the results.`

Line 316: In `Refer the Projects Ideas list` add `to` before `the`.

AyanSinhaMahapatra replied:
Addressed, thanks!

johnmhoran commented on docs/source/how-to-guides/add_new_license.rst:
docs/source/how-to-guides/add_new_license.rst

Line 9: In `To add new license` insert `a` before `new`.

Line 15: In `numbers from 0 to 9,and` add a space after the comma.

Line 18: In `less than 50 characters` change `less` to `fewer`.

Line 20: In "We also have to add a `spdx_license_key`" change `a` to `an`.

Line 24: In `I.e. the filename` change `I.e.` to `For example,`.

Line 30: Change `characters, copyrights` to `characters (e.g., copyrights)`.

Line 33:  In `the data attributes for the license in YAML format as YAML frontmatter.`, what does `as YAML frontmatter` mean?

Line 46: Change `but nice to have always` to `but always nice to have`.

Line 54: Change `which are very similar` to `that are very similar`.

Line 55: Change `make sure we match correctly these licenses` to `make sure we match these licenses correctly`.

Line 79: Change `developement` to `development`.

Line 80: Change `to make sure we reindex the licenses and this validates the new licenses.` to `to make sure we reindex the licenses (and this also validates the new licenses).`

AyanSinhaMahapatra replied:
See https://github.com/eyeseast/python-frontmatter, also added this link here. It's a way of adding metadata to plain text files. We recently added this to half the number of files under `data/licenses/`

Comments addressed, thanks!

johnmhoran commented on docs/source/how-to-guides/add_new_license_detection_rule.rst:
docs/source/how-to-guides/add_new_license_detection_rule.rst

Line 32: Change `this name, for example` to `this name; for example,`.

Line 35: In `Save your rule text in this file, if there are specific words`, change `,` to `;`.

Line 64: Change `Each rules needs have one flag to describe the type of license rule, the options are: to `Each rule needs to have one flag to describe the type of license rule. The options are:`

Line 86: Change `to make sure we reindex the rules and this validates the new licenses.` to `to make sure we reindex the rules (and this also validates the new licenses).`

AyanSinhaMahapatra replied:
Not adding the last parenthesis, as it's not a side effect, that is the reason why we run the reindexing.

Addressed other comments, thanks!

johnmhoran commented on docs/source/how-to-guides/install_new_license_plugin.rst:
docs/source/how-to-guides/install_new_license_plugin.rst

Line 141: In `directory contains a file for each license:` change `a` to `one`.

AyanSinhaMahapatra replied:
Addressed, thanks!

johnmhoran commented on docs/source/index.rst:
docs/source/index.rst

Line 103: Change `about scancode-toolkit, how it works or supported features.` to `about scancode-toolkit, including how it works and supported features.`

AyanSinhaMahapatra replied:
Addressed, thanks!

johnmhoran commented on docs/source/misc/faq.rst:
docs/source/misc/faq.rst

Line 28: Change `(~32,000 and` to `(~32,000) and`.

Line 29: In `as opposed to be regex-driven` delete `be`.

Line 70: In `Yes but and this is fully supported and tested.` delete `but`.

Line 72: In `for a previous bug which was preventing this.` change `which` to `that`.

Line 75: In `is now fixed but and has been extensively tested for ScanCode.` delete `but`.

AyanSinhaMahapatra replied:
Addressed, thanks!

johnmhoran commented on docs/source/misc/support.rst:
docs/source/misc/support.rst

Line 20: Change `features open a` to `features, open a`.

Line 26: Change `with aboutcode community members and if you want to talk` to `with AboutCode community members, and if you want to talk`.

AyanSinhaMahapatra replied:
Addressed, thanks!

johnmhoran commented on docs/source/reference/available_package_parsers.rst:
docs/source/reference/available_package_parsers.rst

Line 8: `Scancode supports` -- let's try to be consistent with our product naming -- at least change this to `ScanCode`, although it's actually `ScanCode Toolkit`, isn't it?

AyanSinhaMahapatra replied:
Chnaged to ScanCode. It's both scancode.io and scancode-toolkit, as scancode.io also uses toolkit and this is also what is supported there.

johnmhoran commented on docs/source/reference/license-detection-reference.rst:
docs/source/reference/license-detection-reference.rst

Line 14: Consider changing `specially` to `especially`.

Line 31: Change `Some more elements are still WIP,` to `Some other elements are still WIP;`.

Line 41: Change `seperated` to `separated`.

Line 43:  It's not clear what we're trying to convey here: `We should be mostly certain of a proper detection to report a LicenseDetection.`

Line 46: Change `but we only add more matches to rectify` to `but we add more matches only to rectify`.

Line 79: Change `it's` to `its`.

Line 104: Change `Chnagelog Summary` to `Changelog Summary`.

Lines 110-113: It's not clear to me what we're trying to say here, but the use of `also` (twice) is confusing: 

The data structure of the JSON output has changed for licenses at resource
level, also with new attribute names, ``licenses`` -> ``license_detections``
and ``license_expressions`` -> ``detected_license_expression`` also with a
SPDX version of the same.

Line 121: Change `it's` to `its`.

Line 127: Delete `instead` -- we already use `Instead` to start the sentence in line 126.

Line 129: Where we say `attributred`, do we mean instead to say `attribute`?  `attributed`?

Line 153: Change `where` to `were`.

Lines 302-303: Where we say `Unique license detection`, do we mean to say `Unique License Detection`?  We use that phrase a handful of times on this RTD page, with many variations on the presence/absence of initial uppercase letters.

Line 363: Change "on an ``license-expression``." to "on a ``license-expression``."

Line 366: Change `there was two entries` to `there were two entries`.

Line 376: Change `incosistencies` to `incosistencies`.

Line 379: Change `as a reference optionally.` to `as an optional reference.`

Line 382: Change `eveolved.` to `evolved.`

Line 498: This capitalization looks odd -- is that truly what we want?  `Only reference License related Data`

Line 685: Change `objcts` to `objects`.

AyanSinhaMahapatra replied:
Addressed, thanks!

johnmhoran commented on docs/source/rst_snippets/basic_options.rst:
docs/source/rst_snippets/basic_options.rst

Line 81: Change `with equal or less number of matches` to `with equal or smaller number of matches`.

AyanSinhaMahapatra replied:
Addressed, thanks!

johnmhoran commented on docs/source/rst_snippets/custom_output_format.rst:
docs/source/rst_snippets/custom_output_format.rst

Line 44: Change "Now are results are saved in ``output.txt``" to "Our results are now saved in ``output.txt``".

AyanSinhaMahapatra replied:
Addressed, thanks!

johnmhoran commented on docs/source/rst_snippets/note_snippets/basic_clpieu.rst:
docs/source/rst_snippets/note_snippets/basic_clpieu.rst

Lines 3-4:  Change `If any of combination of these options are used` to `If any combination of these options is used`.

Line 4: Change `ScanCode only performs that specific task` to `ScanCode performs only that specific task`.

Line 5: Change `only scans` to `scans only`.

Line 6: Change "The only notable exception being ``--package`` scan also has" to "The only notable exception, a ``--package`` scan, also has".

Line 8: Our intended meaning is unclear, but my best guess is that we should change "regardless of ``--license`` option being used" to "regardless of whether the ``--license`` option is used".  If there is more than one such option we intend to refer to here, then replace `whether the` with `which`.

AyanSinhaMahapatra replied:
Addressed, thanks!

johnmhoran commented on docs/source/rst_snippets/note_snippets/post_license_policy.rst:
docs/source/rst_snippets/note_snippets/post_license_policy.rst

Lines 6-9:  The full passage:

We do not have
licenses as an required option because this plugin would be
upgraded to also include the license policy attribute for
packages too.

- Change `an required option` to `a required option`.
- I don't understand what we are trying to say with this portion of the passage:

because this plugin would be
upgraded to also include the license policy attribute for
packages too.

AyanSinhaMahapatra replied:
We could have made the ``--license`` option a required option to run ``--license-policy``, but that's not the case because in the future, this would also include packages.

johnmhoran commented on docs/source/rst_snippets/output_format_options.rst:
docs/source/rst_snippets/output_format_options.rst

Line 22: Change `inputs` to `input`.

AyanSinhaMahapatra replied:
Addressed, thanks!

johnmhoran commented on docs/source/rst_snippets/post_scan_options.rst:
docs/source/rst_snippets/post_scan_options.rst

Line 17: Change `now provide a improved consolidated data.` to `now provide improved consolidated data.`

Lines 39, 53 and 63: Change `occurance` to `occurrence`.

AyanSinhaMahapatra replied:
Addressed, thanks!

johnmhoran commented:
@AyanSinhaMahapatra I've completed my review of the 59 files.

AyanSinhaMahapatra committed:
Address feedback and comments

Adds updates and fixes wrt. feedback and comments provided
by @johnmhoran

Signed-off-by: Ayan Sinha Mahapatra <ayansmahapatra@gmail.com>

AyanSinhaMahapatra commented:
@johnmhoran thanks a lot for the feedback and comments, they were really helpful in making the documentation better, more correct and consistent. Have addressed all comments and kept some answers unresolved above for you to see.
